litex/soc/software/libbase/isr.c

```diff
@@ -191,18 +191,46 @@ void isr(void)
 }
 
 #else
-void isr(void)
+#define NR_IRQ 32
+struct irq_table
 {
-	__attribute__((unused)) unsigned int irqs;
+	void (*isr)(int);
+} irq_table[NR_IRQ];
 
-	irqs = irq_pending() & irq_getmask();
+int irq_attach(unsigned int irq, void (*isr)(int))
+{
+	if (irq >= NR_IRQ) {
+		printf("Inv irq %d\n", irq);
+		return -1;
+	}
 
-#ifdef CSR_UART_BASE
-#ifndef UART_POLLING
-	if(irqs & (1 << UART_INTERRUPT))
-		uart_isr();
-#endif
-#endif
+	unsigned int ie = irq_getie();
+	irq_setie(0);
+	irq_table[irq].isr = isr;
+	irq_setie(ie);
+	return irq;
+}
+
+int irq_detach(unsigned int irq)
+{
+	return irq_attach(irq, NULL);
+}
+
+void isr(void)
```

Comment:
Some considerations:
- Assert if irq within NR_IRQ
- NR_IRQ can be generated in soc.py see https://github.com/enjoy-digital/litex/pull/1815#issuecomment-1784151465